toolkit/cli/smart-contracts-commands/src/permissioned_candidates.rs

```diff
@@ -0,0 +1,51 @@
+use crate::parse_partnerchain_public_keys;
+use crate::PaymentFilePath;
+use ogmios_client::jsonrpsee::client_for_url;
+use partner_chains_cardano_offchain::await_tx::FixedDelayRetries;
+use partner_chains_cardano_offchain::permissioned_candidates::upsert_permissioned_candidates;
+use sidechain_domain::UtxoId;
+use std::fs::read_to_string;
+
+#[derive(Clone, Debug, clap::Parser)]
+pub struct UpsertPermissionedCandidatesCmd {
+	#[clap(flatten)]
+	common_arguments: crate::CommonArguments,
+	/// Path to the file containing the permissioned candidates data.
+	/// Each line represents one permissioned candidate in format SIDECHAIN_KEY:AURA_KEY:GRANDPA_KEY
+	#[arg(long)]
+	permissioned_candidates_file: String,
+	#[clap(flatten)]
+	payment_key_file: PaymentFilePath,
+	#[arg(long, short('c'))]
+	genesis_utxo: UtxoId,
+}
+
+impl UpsertPermissionedCandidatesCmd {
+	pub async fn execute(self) -> crate::CmdResult<()> {
+		let payment_key = self.payment_key_file.read_key()?;
+
+		let mut permissioned_candidates = Vec::new();
+
+		for line in read_to_string(&self.permissioned_candidates_file).expect("Permissioned candidates file with each line representing one candidate").lines() {
+			if line.is_empty() {
+                break;
```

Comment:
I think `continue;` would be a bit better, but I don't expect empty lines in other place than at the end of file